raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -1203,41 +1205,40 @@ public void testLeaderHandlesFindQuorum() throws IOException {
     public void testLeaderGracefulShutdown() throws Exception {
         int otherNodeId = 1;
         Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
-        KafkaRaftClient client = buildClient(voters);
-
-        // Elect ourselves as the leader
-        assertEquals(ElectionState.withVotedCandidate(1, localId, voters), quorumStateStore.readElectionState());
-
-        pollUntilSend(client);
-
-        initializeVoterConnections(client, voters, 1, OptionalInt.empty());
-
-        pollUntilSend(client);
-
-        int voteCorrelationId = assertSentVoteRequest(1, 0, 0L);
-        VoteResponseData voteResponse = voteResponse(true, Optional.empty(), 1);
-        deliverResponse(voteCorrelationId, otherNodeId, voteResponse);
-        client.poll();
-        assertEquals(ElectionState.withElectedLeader(1, localId, voters), quorumStateStore.readElectionState());
+        int epoch = 1;
+        KafkaRaftClient client = initializeAsLeader(voters, epoch);
 
         // Now shutdown
         int shutdownTimeoutMs = 5000;
-        client.shutdown(shutdownTimeoutMs);
+        CompletableFuture<Void> shutdownFuture = client.shutdown(shutdownTimeoutMs);
 
         // We should still be running until we have had a chance to send EndQuorumEpoch
+        assertTrue(client.isShuttingDown());
         assertTrue(client.isRunning());
+        assertFalse(shutdownFuture.isDone());
 
-        // Send EndQuorumEpoch request to the other vote
+        // Send EndQuorumEpoch request to the other voter
         client.poll();
+        assertTrue(client.isShuttingDown());
         assertTrue(client.isRunning());
-
         assertSentEndQuorumEpochRequest(1, OptionalInt.of(localId), otherNodeId);
 
-        // Graceful shutdown completes when the epoch is bumped
-        deliverRequest(voteRequest(2, otherNodeId, 0, 0L));
-
+        // We should still be able to handle vote requests during graceful shutdown
+        // in order to help the new leader get elected
+        deliverRequest(voteRequest(epoch + 1, otherNodeId, epoch, 1L));
         client.poll();
-        assertFalse(client.isRunning());
+        assertSentVoteResponse(Errors.NONE, epoch + 1, OptionalInt.empty(), true);
+
+        // Graceful shutdown completes when a new leader is elected
+        deliverRequest(beginEpochRequest(2, otherNodeId));
+
+        TestUtils.waitForCondition(() -> {
```

Comment:
This test case should be deterministic, maybe consider calling `client.poll` twice?

Reply:
But is the number of polls significant to the test? I would argue not. It makes these test cases more brittle when we assume a specific ordering of events in the event loop. We ran into this problem a lot with the transaction manager tests in the client as well.

Reply:
Ok, makes sense